test/src/unit-capi-dynamic.cc

```diff
@@ -0,0 +1,841 @@
+          test_attr.name_,
+          a_read_buffer,
+          &a_read_buffer_size,
+          a_read_buffer_var,
+          &a_read_buffer_var_size,
+          nullptr,
+          nullptr);
+    } else {
+      read_query_buffers.emplace_back(
+          test_attr.name_,
+          a_read_buffer,
+          &a_read_buffer_size,
+          nullptr,
+          nullptr,
+          nullptr,
+          nullptr);
+    }
+  }
+
+  // Execute a read query over the entire domain.
+  uint64_t* subarray_full =
+      (uint64_t*)malloc(2 * sizeof(test_dims) * sizeof(uint64_t*));
+  for (uint64_t i = 0; i < sizeof(subarray_full); i += 2) {
+    for (auto dims_iter = test_dims.begin(); dims_iter != test_dims.end();
+         dims_iter++) {
+      const uint64_t min_range = ((uint64_t*)(dims_iter->domain_))[0];
+      const uint64_t max_range = ((uint64_t*)(dims_iter->domain_))[1];
+      subarray_full[i] = min_range;
+      subarray_full[i + 1] = max_range;
+      break;
+    }
+  }
+
+  read(array_name, read_query_buffers, subarray_full, encryption_type);
+
+  // Each value in `a_read_buffer` corresponds to its index in
+  // the original `a_write_buffer`. Check that the ordering of
+  // the validity buffer matches the ordering in the value buffer.
+  // Validity buffer is set only when the attribute is nullable.
+  if (test_attr.nullable_) {
+    REQUIRE(a_read_buffer_size == a_write_buffer_size);
+    REQUIRE(a_read_buffer_validity_size == a_write_buffer_validity_size);
+    if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+      uint8_t expected_a_read_buffer_validity[2];
+      REQUIRE(a_read_buffer_var_size == a_write_buffer_var_size);
+      for (int i = 0; i < 2; ++i) {
+        const uint64_t idx = a_read_buffer_var[i];
+        expected_a_read_buffer_validity[i] = a_write_buffer_validity[idx];
+      }
+      REQUIRE(!memcmp(
+          a_read_buffer_validity,
+          expected_a_read_buffer_validity,
+          a_read_buffer_validity_size));
+    } else {
+      uint8_t expected_a_read_buffer_validity[1];
+      expected_a_read_buffer_validity[0] = a_write_buffer_validity[0];
+      REQUIRE(!memcmp(
+          a_read_buffer_validity,
+          expected_a_read_buffer_validity,
+          a_read_buffer_validity_size));
+    }
+
+  } else {
+    REQUIRE(a_read_buffer_size == a_write_buffer_size);
+    if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+      uint8_t expected_a_read_buffer[2];
+      REQUIRE(a_read_buffer_var_size == a_write_buffer_var_size);
+      for (int i = 0; i < 2; ++i) {
+        const uint64_t idx = a_read_buffer_var[i];
+        expected_a_read_buffer[i] = a_write_buffer[idx];
+      }
+      REQUIRE(
+          !memcmp(a_read_buffer, expected_a_read_buffer, a_read_buffer_size));
+    } else {
+      uint8_t expected_a_read_buffer[1];
+      expected_a_read_buffer[0] = a_write_buffer[0];
+      REQUIRE(
+          !memcmp(a_read_buffer, expected_a_read_buffer, a_read_buffer_size));
+    }
+  }
+}
+
```

Comment:
Unlike `a_read_buffer` which is allocated on the stack, the buffers within `d_write_buffer` will leak when this routine exists. Make sure you `free` them once you're done with them.